cs3api4lab/handlers.py

```diff
@@ -2,15 +2,114 @@
+        return Cs3ShareApi(self.log)
+
+    @web.authenticated
+    @gen.coroutine
+    def post(self):
+        request = self.get_json_body()
+        RequestHandler.handle_request(self,
+                                      self.share_api.create,
+                                      201,
+                                      request['endpoint'],
+                                      request['file_path'],
+                                      request['grantee'],
+                                      request['idp'],
+                                      request['role'],
+                                      request['grantee_type'])
+
+    @web.authenticated
+    @gen.coroutine
+    def delete(self):
+        share_id = self.get_query_argument('share_id')
+        RequestHandler.handle_request(self, self.share_api.remove, 204, share_id)
+
+    def put(self):
+        request = self.get_json_body()
+        RequestHandler.handle_request(self,
+                                      self.share_api.update,
+                                      204,
+                                      request['share_id'],
+                                      request['role'])
+
+
+class ListSharesHandler(APIHandler):
+    @property
+    def share_api(self):
+        return Cs3ShareApi(self.log)
 
     @web.authenticated
     @gen.coroutine
     def get(self):
-        output = {
-            'hello': 'world'
-        }
-        self.set_header('Content-Type', 'application/json')
-        self.finish(json.dumps(output))
+        RequestHandler.handle_request(self, self.share_api.list_dir_model, 200)
+
+
+class ListReceivedSharesHandler(APIHandler):
+    @property
+    def share_api(self):
+        return Cs3ShareApi(self.log)
+
+    @web.authenticated
+    @gen.coroutine
+    def get(self):
+        RequestHandler.handle_request(self, self.share_api.list_received, 200)
+
+    @web.authenticated
+    @gen.coroutine
+    def put(self):
+        share_id = self.get_query_argument('share_id', default=None)
+        state = self.get_query_argument('state', default='pending')
+        RequestHandler.handle_request(self, self.share_api.update_received, 200, share_id, state)
+
 
-handlers = [(r"/api/cs3/helloworld", HelloWorldHandle)]
\ No newline at end of file
+class ListSharesForFile(APIHandler):
+    @property
+    def share_api(self):
+        return Cs3ShareApi(self.log)
+
+    @web.authenticated
+    @gen.coroutine
+    def get(self):
+        file_path = self.get_query_argument('file_path')
+        RequestHandler.handle_request(self, self.share_api.list_grantees_for_file, 200, file_path)
+
+
+handlers = [
+    (r"/api/cs3/shares", ShareHandler),
+    (r"/api/cs3/shares/list", ListSharesHandler),
+    (r"/api/cs3/shares/received", ListReceivedSharesHandler),
+    (r"/api/cs3/shares/file", ListSharesForFile),
+]
+
+
+class RequestHandler(APIHandler):
+
+    @staticmethod
+    def handle_request(self, api_function, success_code, *args):
```

Comment:
Your error handling only produces 200 OKs + requested outputs OR 500. Isn't there a middle term?
Like, file doesn't exist or no permission to change the shares or smth?

Reply:
For now let's keep as is and iterate later. But please add a #TODO in the code or an issue in github to followup.

Reply:
I added an issue and will take care of that after this PR. I will add additional codes, however it is worth noting that reva itself returns only 200s and 500s on most of the errors

Reply:
Take note of all these limitations. We should give our findings as a list of improvements to the CS3APIs!